In [1]:
# Import libaries
import pandas as pd
import numpy as np
import os

In [2]:
#Define path
path = r'/Users/shikongo/Instacart Basket Analysis/'

In [3]:
path

'/Users/shikongo/Instacart Basket Analysis/'

In [4]:
df_orders_products_merged = pd.read_pickle(r'/Users/shikongo/Instacart Basket Analysis/02 Data/Prepared Data/orders_products_merged.pkl')

In [5]:
df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))

1. Create a subset of the dataframe

In [6]:
df_orders_products_merged = df_orders_products_merged[:1000000]

In [7]:
df_orders_products_merged.shape

(1000000, 19)

2. Create condition to include filters

In [8]:
# Defined a function
def price_label(row):
    
    if row['prices']<= 5:
        return 'Low-range product'
    elif(row['prices'] < 5) and (row['prices']<= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: return 'Not enough data'

In [9]:
# Apply the function
df_orders_products_merged['price_range'] = df_orders_products_merged.apply(price_label, axis=1)

In [10]:
df_orders_products_merged['price_range'].value_counts(dropna = False)

Not enough data      659373
Low-range product    331283
High range             9344
Name: price_range, dtype: int64

3.Check most expensive product

In [11]:
df_orders_products_merged['prices'].max()

24.5

4.Creating condition if-else use loc.

In [13]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
df_orders_products_merged.loc[(df_orders_products_merged['prices'] <= 15) & (df_orders_products_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [15]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
df_orders_products_merged['price_range'].value_counts(dropna = False)

Not enough data      659373
Low-range product    331283
High range             9344
Name: price_range, dtype: int64

In [17]:
# find day most orders take place
df_orders_products_merged['orders_day_of_wek'].value_counts(dropna = False)

0    184034
1    174567
6    136739
2    132114
5    131867
3    120940
4    119739
Name: orders_day_of_wek, dtype: int64

5. Print the frequency of the new column

In [18]:
result = []

for value in df_orders_products_merged["orders_day_of_wek"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")


In [19]:
# Check the result
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

6. Rename the "Busiest day" column and identify the two busiest and two slowest days

In [20]:
df_orders_products_merged.rename(columns = {'Busiest day' : 'Busiest days'}, inplace = True)

In [21]:
df_orders_products_merged['orders_day_of_wek'].value_counts(dropna = False) 

0    184034
1    174567
6    136739
2    132114
5    131867
3    120940
4    119739
Name: orders_day_of_wek, dtype: int64

7. Print the frequency of the new column

In [22]:
result = []

for value in df_orders_products_merged["orders_day_of_wek"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 3 or value == 4:
    result.append("slowest days")
  else:
    result.append("Regular days")

In [23]:
result

['Regular days',
 'Regular days',
 'Busiest days',
 'slowest days',
 'slowest days',
 'Busiest days',
 'Regular days',
 'slowest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'slowest days',
 'slowest days',
 'Regular days',
 'slowest days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'slowest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'slowest days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days

8.Create a new column for the busiest and slowest days

In [24]:
df_orders_products_merged['Busiest_days'] = result

In [25]:
df_orders_products_merged['Busiest_days'].value_counts(dropna = False)

Regular days    400720
Busiest days    358601
slowest days    240679
Name: Busiest_days, dtype: int64


Observations:

"Regular days" have the highest count: This category has the most occurrences, with 400,720. This suggests that a significant number of days fall into this category.

"Busiest days" have the second-highest count: "Busiest days" are close in count to "Regular days," with 358,601 occurrences. This indicates that there are periods when order activity is high, but they are not as frequent as regular days.

"Slowest days" have the lowest count: "Slowest days" have the fewest occurrences, with 240,679. This implies that there are specific days or periods with relatively low order activity.


9.Show the hours and frequencies

In [26]:
df_orders_products_merged['order_hour_of_day'].value_counts(dropna = False)

10    86399
11    84804
14    82985
13    82197
15    81121
12    80503
16    77638
9     77082
17    64025
8     53631
18    49977
19    38479
20    30380
7     27652
21    24383
22    19139
23    12126
6      9105
0      6540
1      3550
5      2847
2      2173
3      1639
4      1625
Name: order_hour_of_day, dtype: int64

10.Print the frequency of the new column

In [27]:
result = []

for value in df_orders_products_merged["order_hour_of_day"]:
  if value >= 9 and value <=16:
    result.append("Most_orders")
  elif value >= 0 and value <=6:
    result.append("Fewest_orders")
  else:
    result.append("Average_orders")

In [28]:
df_orders_products_merged['Busiest_period_of_day'] = result

In [29]:
result

['Most_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Fewest_orders',
 'Average_orders',
 'Fewest_orders',
 'Fewest_orders',
 'Average_orders',
 'Fewest_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Fewest_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Most_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_or

11.Create column for the most orders, average orders and fewest orders

In [30]:
df_orders_products_merged['Busiest_period_of_day'] = result

In [31]:
df_orders_products_merged['Busiest_period_of_day'].value_counts(dropna = False)

Most_orders       652729
Average_orders    319792
Fewest_orders      27479
Name: Busiest_period_of_day, dtype: int64

In [32]:
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_update_merged.pkl'))

In [33]:
df_orders_products_merged.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   

  order_hour_of_day  days_since_prior_order  ordr_id  add_to_cart_order  \
0                11                     3.0  3139998                  5   
1                17                    20.0  1977647                  1   
2                21                     6.0   389851                 20   
3                13                     NaN   652770                 10   
4                17                     9.0  1813452                 11   

   reordered  _merge      price_range    price_range_loc  Busiest_days  \
0          0    both  Not enough data  Mid-range product  Regular days   
1          1    both  Not enough data  Mid-range product  Regular days   
2          0    both  Not enough data  Mid-range product  Busiest days   
3          0    both  Not enough data  Mid-range product  slowest days   
4          1    both  Not enough data  Mid-range product  slowest days   

  Busiest_period_of_day  
0           Most_orders  
1        Average_orders  
2        Average_orders  
3           Most_orders  
4        Average_orders  

[5 rows x 23 columns]